src/steps/os/unix.rs

```diff
@@ -432,6 +403,123 @@ pub fn run_nix(ctx: &ExecutionContext) -> Result<()> {
     }
 }
 
+pub fn run_nix_self_upgrade(ctx: &ExecutionContext) -> Result<()> {
+    let nix = require("nix")?;
+
+    // Should we attempt to upgrade Nix with `nix upgrade-nix`?
+    #[allow(unused_mut)]
+    let mut should_self_upgrade = cfg!(target_os = "macos");
+
+    #[cfg(target_os = "linux")]
+    {
+        // We can't use `nix upgrade-nix` on NixOS.
+        if let Ok(Distribution::NixOS) = Distribution::detect() {
+            should_self_upgrade = false;
+        }
+    }
```

Comment:
```suggestion
    // Should we attempt to upgrade Nix with `nix upgrade-nix`?
    let mut should_self_upgrade = if cfg!(target_os = "macos") {
        true
    } else if cfg!(target_os = "linux") {
        !matches!(Distribution::detect(), Distribution::NixOS)
    } else {
        unreachable!()
    };
```

Maybe we can do this here, it has better readability, though the generated code looks bad:

```rs
// On Linux
    let should = if false {
        true
    } else if true {
        match Distribution::detect() {
            Distribution::NixOS => true,
            _ => false,
        }
    } else {
        ::core::panicking::panic("internal error: entered unreachable code")
    };

```

Reply:
Ah, I am sorry, I didn't realize that this `Distribution` type is only defined on Linux, and since `cfg!(target_os = "macos")` will be simply translated to a bool value, it requires that this type is also available on macOS, so it won't compile:(

Reply:
Ahh, I'll revert.